### PR TITLE
Add TimesFM 3 multivariate forecasting support

### PR DESCRIPTION
**Pull request by justinchuby**
## Summary

- add a native declarative ONNX implementation of [TimesFM 3](https://research.google/blog/timesfm-3-a-zero-shot-foundation-model-for-multivariate-forecasting/)
- export a five-component tensor-only pipeline:
  1. `raw_preprocessor`: padded-batch interpolation, dynamic padding, detrending, role/mask construction, and patching
  2. `preprocessor`: running RevIN and patched feature construction
  3. `model`: every learned layer (residual projection, temporal/variate transformer, output head), with no `Scan`, `Loop`, or `If`, so it can be captured independently
  4. `postprocessor`: iterative CPM RevIN refinement and reverse normalization
  5. `stitcher`: overlap stitching, trend reversal, quantile sorting, optional positivity, and padded validity masks
- support heterogeneous context and horizon lengths in one padded batch
- recognize the official `google/timesfm-3.0-pytorch` config, which does not declare a Transformers `model_type`
- add `examples/timesfm3_forecasting.py`, a reusable five-session driver with CUDA I/O binding, fixed-shape CUDA graph capture, latency reporting, CSV output, and official PyTorch parity/benchmark comparison

The implementation follows `google-research/timesfm` commit `331c6d33cb1ac2611de3056d0ac7164aab6301eb` and checkpoint revision `900fcab43d1bfe71733a33b3fec61a41fce28a27`.

## Input contract

`raw_preprocessor` accepts right-aligned context tensors `[B,V,C]`, left-aligned future-covariate tensors `[B,V,H]`, observed-value masks, per-row context/horizon lengths, and roles `0=target`, `1=past-only`, `2=past-future`. Outputs remain padded and include explicit validity/target masks; no ragged tensor convention is required.

Symmetric averaging and outer z-normalization remain opt-in host orchestration policies because they duplicate/remap whole requests rather than form part of the checkpoint forward pass.

## Validation

- exact tiny PyTorch/ONNX parity across preprocessing, temporal attention, variate attention, and CPM refinement: max absolute error about `1.1e-6`
- real official-weight AirPassengers forecast matches the upstream PyTorch forecaster with max absolute difference `9.2e-5`
- AirPassengers 24-month holdout: TimesFM MAE `36.1` versus seasonal-naive MAE `71.3`
- official 20-layer config builds and all 445 learned parameters route uniquely to the capture-friendly `model` component
- FP32/FP16 patched execution, raw interpolation/detrending, heterogeneous padding invariance, per-row stitching, quantile sorting, and positivity policy covered by focused ORT tests
- targeted model suite: 10 passed
- transformer config/builder/coverage regressions: 861 passed, 277 skipped
- full non-integration baseline before the component split: 9,850 passed, 68 skipped; 17 unrelated existing/environment-dependent failures outside the changed surfaces

### CUDA benchmark

Measured on an NVIDIA RTX 4060 Laptop GPU with ONNX Runtime 1.29.0 and PyTorch 2.11.0+cu128, using the official 330M checkpoint, batch 1, one variate, 512 context points, 64 forecast points, and 50 steady-state learned-core iterations:

| Runtime | Precision | Core p50 | Notes |
|---|---:|---:|---|
| ONNX Runtime CUDA graph | FP16 | 105.9 ms | Fixed-shape I/O binding; no host/device copies in core timing |
| ONNX Runtime CUDA | FP16 | 133.6 ms | Same-process comparison without graph capture |
| Official PyTorch CUDA | FP32 | 147.0 ms | 10 iterations; upstream FP16 RoPE promotes Q/K but not V, which PyTorch SDPA rejects |

CUDA graph capture reduced ORT learned-core median latency by about 21% in the controlled same-process comparison. FP16 ONNX versus FP32 PyTorch is not a like-for-like precision comparison; the example reports this fallback explicitly and still checks outputs (learned-core max absolute difference `0.00843`, final forecast `0.00926`).

## Licensing and waivers

The upstream implementation is Apache-2.0. The official TimesFM 3 weights use Google's non-commercial license, so this PR does not include, redistribute, cache, or download those weights in CI.

**Waivers:**
- L2/L4/L5 CI goldens and real-weight integration are not committed because they would require CI to download the non-commercial 1.3 GB checkpoint. Real-weight parity was run locally against the pinned revision instead.
- ORT GenAI and text-generation metadata do not apply to the time-series forecasting pipeline.

### CODE REVIEW

File: src/mobius/tasks/_time_series_forecasting.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    """
+
+    model_roles: ClassVar[dict[str, str]] = {"model": "encoder"}
+
+    def build(self, module: nn.Module, config: BaseModelConfig) -> ModelPackage:
+        batch = ir.SymbolicDim("batch")
+        variates = ir.SymbolicDim("variates")
+        patches = ir.SymbolicDim("patches")
+        patch_len = config.input_patch_len
+
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/tasks/_time_series_forecasting.py:51]: Fixed in bbcb2cab. The task now reads `input_patch_len` through `getattr`, validates that it is an integer (excluding bool), and raises a clear `TypeError` for incompatible configs. Added a regression test using `BaseModelConfig`.

File: src/mobius/models/timesfm3.py [hunk 2]
```diff
@@ -0,0 +1,787 @@
...
+    def validate(self) -> None:
+        if self.output_patch_len % self.input_patch_len:
+            raise ValueError("output_patch_len must be a multiple of input_patch_len")
+        if self.model_dims % self.num_heads:
+            raise ValueError("model_dims must be divisible by num_heads")
+        if self.head_dim % 2:
+            raise ValueError("TimesFM 3 RoPE requires an even head_dim")
+        if not self.quantiles:
+            raise ValueError("quantiles must not be empty")
+
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/timesfm3.py:89]: Fixed in bbcb2cab. `TimesFM3Config.validate()` now rejects non-integer, boolean, zero, and negative input/output patch lengths before modulo or graph construction. Added regression coverage for zero and negative lengths.

File: src/mobius/integrations/transformers/_config_resolver.py [hunk 3]
```diff
@@ -136,6 +139,24 @@ def _try_load_config_json(model_id: str, revision: str | None = None):
     return _dict_to_pretrained_config(config_dict)
 
 
+def _model_type_from_config_signature(config: dict) -> str | None:
```
Comment (justinchuby) [inline src/mobius/integrations/transformers/_config_resolver.py:142]: This seems overly complicated. We can just read the google/timesfm-3.0-pytorch model name.
